test/libyul/yulOptimizerTests/unusedStoreEliminator/mcopy_sstore.yul

```diff
@@ -0,0 +1,32 @@
+{
+    calldatacopy(0, 0, 0x40)
+
+    let _0 := 0
+    let _32 := 32
+    let _42 := 42
+    let _123 := 123
+
+    // Sanity check: MCOPY does not affect storage
+    sstore(_0, _42) // Redundant. SSTORE overwrites it.
+    mcopy(_32, _0, _32)
+    sstore(_0, _123)
+
+    return(0, 0x40)
+}
+// ====
+// EVMVersion: >=cancun
+// ----
+// step: unusedStoreEliminator
+//
+// {
+//     {
+//         calldatacopy(0, 0, 0x40)
+//         let _0 := 0
+//         let _32 := 32
+//         let _42 := 42
+//         let _123 := 123
+//         mcopy(_32, _0, _32)
+//         sstore(_0, _123)
+//         return(0, 0x40)
+//     }
+// }
```

Comment:
The annoying thing here is that I had to manually create variables for constants here despite the fact that we run `ExpressionSplitter` and `SSATransform` in the test framework just before testing the step. Still, without it the first `sstore` was not being removed and it makes me worried that it would be interfering with other `UnusedStoreEliminator` tests I added.

The way it is now, the code gets transformed into this just before we run `UnusedStoreEliminator`:
```yul
    {
        let _1 := 0x40
        let _2 := 0
        let _3 := 0
        calldatacopy(_3, _2, _1)
        let _0 := 0
        let _32 := 32
        let _42 := 42
        let _123 := 123
        sstore(_0, _42)
        mcopy(_32, _0, _32)
        sstore(_0, _123)
        let _4 := 0x40
        let _5 := 0
        return(_5, _4)
    }
```
and the first `sstore` gets removed by the step.

but when the input looked like this:

```yul
    ...
    sstore(0, 42)
    mcopy(32, 0, 32)
    sstore(0, 123)
    ...
```

it would instead end up like this before `UnusedStoreEliminator`:
```yul
    {
        let _1 := 0x40
        let _2 := 0
        let _3 := 0
        calldatacopy(_3, _2, _1)
        let _4 := 42
        let _5 := 0
        sstore(_5, _4)
        let _6 := 32
        let _7 := 0
        let _8 := 32
        mcopy(_8, _7, _6)
        let _9 := 123
        let _10 := 0
        sstore(_10, _9)
        let _11 := 0x40
        let _12 := 0
        return(_12, _11)
    }
```
and both `sstore`s would remain.

Unfortunately this means that these tests ended up being more verbose than they could be otherwise.

Reply:
Running `CommonSubexpressionEliminator` before `UnusedStoredEliminator` will get rid of the redundant `sstore` by the way. Obviously, injecting this into the test suite will likely yield changes in a lot of tests, which is not ideal, but we do indeed run CSE multiple times before the `UnusedStoreEliminator` . I'll try to explain, although take everything I say with a grain of salt until Daniel (or maybe even Chris) chime in. The following is the above Yul code transformed after `ExpressionSplitter` and `SSATransform` are run:
```
{
    let _1 := 0x40
    let _2 := 0
    let _3 := 0
    calldatacopy(_3, _2, _1)
    let _4 := 42
    let _5 := 0
    sstore(_5, _4)
    let _6 := 32
    let _7 := 0
    let _8 := 32
    mcopy(_8, _7, _6)
    let _9 := 123
    let _10 := 0
    sstore(_10, _9)
}
```
Only pay attention to `_2` and the destination for `sstore` writes, in the above case, `_5` for the first one, and `_10` for the second. After `CommonSubexpressionEliminator` is run, we get the following:
```
{
    let _1 := 0x40
    let _2 := 0
    let _3 := _2
    calldatacopy(_2, _2, _1)
    let _4 := 42
    let _5 := _2
    sstore(_2, _4)
    let _6 := 32
    let _7 := _2
    let _8 := _6
    mcopy(_6, _2, _6)
    let _9 := 123
    let _10 := _2
    sstore(_2, _9)
}
```
Notice that destinations for both `sstore`s have changed from `_5` and `_10` to `_2`. `UnusedStoreEliminator` checks whether all code paths revert in order to remove an `sstore` (not relevant in this case), and whether all code paths lead to the first `sstore` being overwritten by a subsequent one (definitely relevant in this case). As you can see (after running the CSE), the second `sstore` will clearly overwrite the first `sstore` (`_2`), which is why it will then remove the first one as it's marked as redundant. This is not possible (in this case) without running the CSE, since without it, the `UnusedStoreEliminator` has no way to know whether `_5` and `_10` are the same destination (technically it does, but that's CSE's job).

Having written this up, I'm now fairly surprised we're not running the `CommonSubexpressionEliminator` for `UnusedEliminatorTests` (nor are we suggesting it in the docs).

Reply:
Now that I think about it, the above could use at least the `UnusedPruner` to clean it up a bit, so I guess the assumption for `UnusedStoreEliminator` tests is that the provided code is already in a 'post CSE` form (which is a terrible assumption to be honest, since you have to do so manually as you've done here). All in all, it's gonna make review this incredibly annoying :)

Reply:
> (which is a terrible assumption to be honest, since you have to do so manually as you've done here)

On the other hand these test suites are meant to test a single step. So cramming too many steps into it, especially the more complex ones, might not be a good idea. You're always risking that it will hide a bug. Not sure about CSE specifically - may or may not be acceptable. We already run quite a few others before and after it.

In any case, it sounds like CSE should indeed be listed in the docs as a soft dependency. We should confirm with Daniel it makes sense and add it.

> All in all, it's gonna make review this incredibly annoying :)

I mean, it's mostly annoying due to the extra verbosity. The test themselves are still pretty simple and you can pretty much ignore the declarations when reading since the names reflect the values.

Reply:
> On the other hand these test suites are meant to test a single step. So cramming too many steps into it, especially the more complex ones, might not be a good idea. You're always risking that it will hide a bug. Not sure about CSE specifically - may or may not be acceptable. We already run quite a few others before and after it.

Yup, that's why  left the second comment after giving it some thought.

> In any case, it sounds like CSE should indeed be listed in the docs as a soft dependency. We should confirm with Daniel it makes sense and add it.

Yeah, or maybe we can get some input from @chriseth if he's not off as well?